File: lib/meta.ts
import {
  mkdirSync,
  readFileSync,
  renameSync,
  writeFileSync,
  existsSync,
  unlinkSync,
} from "node:fs";
import { basename, join } from "node:path";
import { EventEmitter } from "node:events";
import type { TaskStatus, TaskSection } from "./tasks";

/**
 * Write `meta.json` atomically: stage the new contents in a sibling
 * tempfile, then `rename` over the destination. `rename` is atomic on
 * POSIX and atomic-on-success on NTFS, so a crash mid-write leaves
 * either the old file or the new one — never a half-written one. The
 * temp file is also a per-call random suffix so two concurrent writes
 * can't trample each other's staging.
 */
function atomicWriteJson(filePath: string, value: unknown): void {
  const dir = filePath.slice(0, Math.max(filePath.lastIndexOf("/"), filePath.lastIndexOf("\\")));
  mkdirSync(dir, { recursive: true });
  const tmp = `${filePath}.${process.pid}.${Date.now()}.${Math.random().toString(36).slice(2, 8)}.tmp`;
  writeFileSync(tmp, JSON.stringify(value, null, 2) + "\n");
  try {
    renameSync(tmp, filePath);
  } catch (err) {
    // Best-effort cleanup of the staged file before re-throwing.
    try { unlinkSync(tmp); } catch { /* ignore */ }
    throw err;
  }
}

export type RunStatus = "queued" | "running" | "done" | "failed" | "stale";

/**
 * A run is a single Claude Code session the coordinator spawned for a task.
 * `role` is an arbitrary label chosen by the coordinator ("coordinator",
 * "coder", "reviewer", "researcher", "fixer", ...) — not a closed enum.
 * `repo` is the registered app name (from `~/.claude/bridge.json`) the
 * session runs against; for orchestration runs it's the bridge folder's
 * own basename (`BRIDGE_FOLDER` from `lib/paths.ts`).
 * `parentSessionId` is the coordinator session UUID that spawned this run
 * via `POST /api/tasks/<id>/agents`. `null` / absent means the run was
 * spawned directly by the bridge (the coordinator itself). The agent
 * tree visualizer uses this to render parent / child relationships.
 */
export interface Run {
  sessionId: string;
  role: string;
  repo: string;
  status: RunStatus;
  startedAt: string | null;
  endedAt: string | null;
  parentSessionId?: string | null;
  /**
   * Phase D auto-retry: when the bridge auto-spawns a fix agent because
   * a child failed, the new run carries the failed run's sessionId here
   * so the AgentTree can group siblings under the same parent visually.
   * `null` / absent means this run is not a retry.
   */
  retryOf?: string | null;
  /**
   * P2 verify chain: when an app has at least one `verify` command
   * configured (`bridge.json.apps[].verify`), the bridge runs the chain
   * after the child exits cleanly and persists the per-step results
   * here. Absent / null = chain didn't run (no commands configured, or
   * the run was a coordinator / unregistered repo).
   */
  verify?: RunVerify | null;
  /**
   * P2b-1 inline verifier: after the verify chain passes, the bridge
   * compares the child's `reports/<role>-<repo>.md` claims against the
   * actual `git diff` to catch agent hallucinations (claimed file
   * changes that didn't happen, or changes the agent didn't mention).
   * Absent / null = verifier didn't run (no report file, no diff to
   * compare, or the run was a retry-of-retry).
   */
  verifier?: RunVerifier | null;
  /**
   * P2b-2 agent-driven style critic: when `app.quality.critic` is on,
   * after the inline verifier passes the bridge spawns a `style-critic`
   * agent to judge whether the diff matches the codebase fingerprint /
   * symbol index / house-rules. `alien` blocks the auto-commit and
   * triggers a `-stretry` retry. Absent / null = gate didn't run.
   */
  styleCritic?: RunStyleCritic | null;
  /**
   * P2b-2 agent-driven semantic verifier: when `app.quality.verifier`
   * is on, after the style critic passes the bridge spawns a
   * `semantic-verifier` agent to judge whether the changes actually
   * accomplish the task body (not just whether the file list matches).
   * `broken` blocks the auto-commit and triggers a `-svretry` retry.
   * Absent / null = gate didn't run.
   */
  semanticVerifier?: RunSemanticVerifier | null;
  /**
   * P4/F1 worktree mode: absolute path of the private git worktree this
   * run executed in. Set when `app.git.worktreeMode === "enabled"` and
   * the bridge created the worktree before spawn. Used by the diff
   * endpoint and the post-exit cleanup. Absent / null = the run used
   * the live tree (legacy / opt-out behavior).
   */
  worktreePath?: string | null;
  /**
   * P4/F1 branch the worktree was created on (the agent's working
   * branch). Persisted so post-exit cleanup can merge it back into
   * `worktreeBaseBranch`. Absent when `worktreePath` is absent.
   */
  worktreeBranch?: string | null;
  /**
   * P4/F1 base branch the worktree forked from. Cleanup merges the
   * agent's branch back into this one. `null` means the worktree was
   * created on a detached HEAD or fresh repo — cleanup will skip the
   * merge and just remove the worktree.
   */
  worktreeBaseBranch?: string | null;
}

/**
 * P2 — outcome of the post-run verify chain. Steps are persisted in
 * declaration order (`format → lint → typecheck → test → build`), but
 * only the steps the app actually configured a command for appear here
 * (no skipped placeholder rows, to keep meta.json terse).
 */
export interface RunVerify {
  steps: RunVerifyStep[];
  /** True iff every step in `steps` exited 0. */
  passed: boolean;
  startedAt: string;
  endedAt: string;
  /** True if the bridge spawned a `-vretry` follow-up because of this. */
  retryScheduled?: boolean;
}

export interface RunVerifyStep {
  /** Canonical step name — matches the AppVerify field key. */
  name: "format" | "lint" | "typecheck" | "test" | "build";
  /** Exact command line the bridge ran via `sh -c` / `cmd /c`. */
  cmd: string;
  ok: boolean;
  /** Process exit code; `null` when the chain aborted (timeout, spawn error). */
  exitCode: number | null;
  durationMs: number;
  /**
   * Combined stdout+stderr, capped at the runner's outputCapBytes
   * (default 16 KB). Truncation marker appended when capped.
   */
  output: string;
}

/**
 * P2b-1 — outcome of the inline claim-vs-diff verifier.
 *
 * Verdicts:
 *   - `pass`     — claimed files match actual diff (or both are empty
 *                  for a read-only run)
 *   - `drift`    — meaningful mismatch: agent claimed files it didn't
 *                  touch, OR touched files it didn't mention. Triggers
 *                  a `-cretry` (claim-retry) follow-up.
 *   - `broken`   — agent claimed changes but the diff is empty (or
 *                  vice-versa: huge diff with empty claims). Strongest
 *                  signal of hallucination; same retry path as drift.
 *   - `skipped`  — verifier didn't run because preconditions weren't
 *                  met (no report file, no git repo, role is itself a
 *                  retry, etc.). NOT a failure — commit proceeds.
 */
export interface RunVerifier {
  verdict: "pass" | "drift" | "broken" | "skipped";
  /** One-line human reason behind the verdict. */
  reason: string;
  /** File paths the child's report listed under `## Changed files`. */
  claimedFiles: string[];
  /**
   * File paths actually touched per `git diff --name-only`. Excludes
   * lockfiles / generated dirs by default — see verifier.ts filter.
   */
  actualFiles: string[];
  /**
   * Files claimed but not in the diff (likely hallucination — agent
   * said it changed X but didn't).
   */
  unmatchedClaims: string[];
  /**
   * Files in the diff but not claimed (likely sloppy reporting —
   * filtered to non-trivial paths).
   */
  unclaimedActual: string[];
  durationMs: number;
  retryScheduled?: boolean;
}

/**
 * P2b-2 — outcome of the agent-driven style critic.
 *
 * Verdicts:
 *   - `match`   — diff fits the codebase's auto-detected fingerprint,
 *                 reuses the helpers it should, follows house-rules.
 *   - `drift`   — minor deviations: surfaced but commit proceeds.
 *   - `alien`   — looks foreign to this codebase. Triggers a `-stretry`
 *                 (style-retry) follow-up; commit is blocked.
 *   - `skipped` — gate didn't run (critic crashed, verdict file missing,
 *                 role exempt). Commit proceeds untouched.
 */
export interface RunStyleCritic {
  verdict: "match" | "drift" | "alien" | "skipped";
  reason: string;
  /** Specific deviations the critic flagged (max ~10 surfaced). */
  issues: string[];
  /** sessionId of the spawned `style-critic` agent — for UI cross-ref. */
  criticSessionId?: string | null;
  durationMs: number;
  retryScheduled?: boolean;
}

/**
 * P2b-2 — outcome of the agent-driven semantic verifier. Distinct from
 * the inline `RunVerifier` (claim-vs-diff) — this one judges INTENT, not
 * file-list honesty.
 *
 * Verdicts mirror the inline verifier shape so the UI can reuse styling:
 *   - `pass`    — changes accomplish the task body; commit proceeds.
 *   - `drift`   — partial / off-target progress; commit proceeds with note.
 *   - `broken`  — clearly does not accomplish the task. Triggers a
 *                 `-svretry` follow-up; commit blocked.
 *   - `skipped` — gate didn't run. Commit proceeds untouched.
 */
export interface RunSemanticVerifier {
  verdict: "pass" | "drift" | "broken" | "skipped";
  reason: string;
  /** Specific gaps the verifier identified (max ~10 surfaced). */
  concerns: string[];
  /** sessionId of the spawned `semantic-verifier` agent. */
  verifierSessionId?: string | null;
  durationMs: number;
  retryScheduled?: boolean;
}

/**
 * Per-task runtime state. The full task definition (title, body, status,
 * section, checked) lives here too — meta.json is the source of truth
 * for tasks, replacing the old `tasks.md` round-trip.
 */
export interface Meta {
  taskId: string;
  taskTitle: string;
  taskBody: string;
  taskStatus: TaskStatus;
  taskSection: TaskSection;
  taskChecked: boolean;
  /**
   * Target app name from the apps registry (`sessions/init.md`).
   * `null` / absent means "auto" — let the coordinator's heuristic
   * decide which repo to dispatch to based on the task body.
   */
  taskApp?: string | null;
  createdAt: string;
  runs: Run[];
}

const FILE = "meta.json";

/**
 * Lifecycle event payload emitted whenever a task's meta.json mutates
 * via this module's helpers. Phase C's per-task SSE route subscribes to
 * this and forwards `spawned` / `done` / `failed` events to the UI so
 * the polling fallback can drop from 1.5s → 5s.
 *
 * - `kind`:
 *   - `"spawned"` — appendRun: brand-new run (`prevStatus === undefined`)
 *   - `"transition"` — updateRun: status changed (prev → next)
 *   - `"updated"` — updateRun without a status change (e.g. role rename)
 *   - `"writeMeta"` — full writeMeta (no per-run signal — task header changed)
 *   - `"retried"` — Phase D: auto-retry kicked off for a failed child;
 *     `run` is the NEW retry run, `retryOf` is the failed run's sessionId
 */
export interface MetaChangeEvent {
  taskId: string;
  kind: "spawned" | "transition" | "updated" | "writeMeta" | "retried";
  sessionId?: string;
  run?: Run;
  prevStatus?: RunStatus;
  retryOf?: string;
}

interface MetaEvents {
  emitter: EventEmitter;
}

// HMR-safe global stash, same trick as permissionStore / spawnRegistry.
const G = globalThis as unknown as { __bridgeMetaEvents?: MetaEvents };
const events: MetaEvents =
  G.__bridgeMetaEvents ?? { emitter: (() => { const e = new EventEmitter(); e.setMaxListeners(0); return e; })() };
G.__bridgeMetaEvents = events;

export function subscribeMeta(
  taskId: string,
  cb: (ev: MetaChangeEvent) => void,
): () => void {
  const handler = (ev: MetaChangeEvent) => {
    if (ev.taskId === taskId) cb(ev);
  };
  events.emitter.on("meta:changed", handler);
  return () => events.emitter.off("meta:changed", handler);
}

/**
 * Resolve a task id from the sessions directory we were handed. Every
 * call site uses `join(SESSIONS_DIR, taskId)`, so the basename is the
 * task id. Cheap and avoids threading the id through every helper.
 */
function taskIdFromDir(dir: string): string {
  return basename(dir);
}

function emit(ev: MetaChangeEvent): void {
  // setImmediate-equivalent: emit synchronously so subscribers see the
  // mutation in the same tick. SSE consumers buffer their own writes.
  events.emitter.emit("meta:changed", ev);
}

/**
 * CRIT-2: per-task-directory async mutex around the read-modify-write
 * helpers below (`appendRun`, `updateRun`, `applyManyRuns`). Without it,
 * two concurrent callers (e.g. coordinator spawning two children, or a
 * `link` POST racing the lifecycle hook) both observe the same
 * pre-mutation file, mutate, and atomically rename — the second write
 * silently overwrites the first, losing one run.
 *
 * The lock is keyed by absolute task dir. The atomic file rename inside
 * each helper still protects against partial writes; the lock just
 * serializes the read window so back-to-back writes can't trample each
 * other.
 *
 * Map-of-promises pattern: every call chains onto whatever the latest
 * pending operation for that dir is. The chain catches errors so a
 * failed op doesn't poison the queue for everyone behind it. We
 * best-effort GC the entry when the chain settles, but a tiny leak is
 * preferable to dropping the chain link mid-flight.
 *
 * Exported so other modules (e.g. `tasksStore.updateTask`,
 * `tasksStore.migrateTaskApp`) can serialize their own `readMeta →
 * mutate header → writeMeta` sequences against the run-row helpers
 * below. Without this, a UI title edit racing a child's appendRun
 * could silently drop the just-appended run.
 */
const writeQueues = new Map<string, Promise<unknown>>();
export async function withTaskLock<T>(dir: string, fn: () => T | Promise<T>): Promise<T> {
  const prev = writeQueues.get(dir) ?? Promise.resolve();
  // Run regardless of whether `prev` resolved or rejected — one bad
  // operation must not strand later writes for the same task.
  const next: Promise<T> = prev.then(
    () => fn(),
    () => fn(),
  );
  // Stash a swallowed-error variant so the queue head can be awaited
  // without re-throwing prior failures into unrelated callers.
  const tail = next.catch(() => {});
  writeQueues.set(dir, tail);
  try {
    return await next;
  } finally {
    if (writeQueues.get(dir) === tail) writeQueues.delete(dir);
  }
}

/**
 * Phase D: fire a `retried` event on the per-task SSE stream after the
 * bridge auto-spawns a fix agent. The `appendRun` call for the new
 * retry run already fires `spawned`; this is a follow-up event the UI
 * can use to draw the retry-of arrow without scanning every run.
 */
export function emitRetried(taskId: string, retryRun: Run, retryOf: string): void {
  emit({
    taskId,
    kind: "retried",
    sessionId: retryRun.sessionId,
    run: retryRun,
    retryOf,
  });
}

export function createMeta(dir: string, header: Omit<Meta, "runs">): void {
  mkdirSync(dir, { recursive: true });
  const meta: Meta = { ...header, runs: [] };
  atomicWriteJson(join(dir, FILE), meta);
  emit({ taskId: taskIdFromDir(dir), kind: "writeMeta" });
}

export function readMeta(dir: string): Meta | null {
  const p = join(dir, FILE);
  if (!existsSync(p)) return null;
  return JSON.parse(readFileSync(p, "utf8")) as Meta;
}

export function writeMeta(dir: string, meta: Meta): void {
  mkdirSync(dir, { recursive: true });
  atomicWriteJson(join(dir, FILE), meta);
  emit({ taskId: taskIdFromDir(dir), kind: "writeMeta" });
}

export async function appendRun(dir: string, run: Run): Promise<void> {
  await withTaskLock(dir, () => {
    const meta = readMeta(dir);
    if (!meta) throw new Error(`meta.json missing at ${dir}`);
    meta.runs.push(run);
    atomicWriteJson(join(dir, FILE), meta);
    emit({
      taskId: taskIdFromDir(dir),
      kind: "spawned",
      sessionId: run.sessionId,
      run,
    });
  });
}

export async function updateRun(
  dir: string,
  sessionId: string,
  patch: Partial<Run>,
): Promise<void> {
  await withTaskLock(dir, () => {
    const meta = readMeta(dir);
    if (!meta) throw new Error(`meta.json missing at ${dir}`);
    const run = meta.runs.find((r) => r.sessionId === sessionId);
    if (!run) throw new Error(`run ${sessionId} not found`);
    const prevStatus = run.status;
    Object.assign(run, patch);
    atomicWriteJson(join(dir, FILE), meta);
    const statusChanged = patch.status !== undefined && patch.status !== prevStatus;
    emit({
      taskId: taskIdFromDir(dir),
      kind: statusChanged ? "transition" : "updated",
      sessionId,
      run: { ...run },
      prevStatus,
    });
  });
}

/**
 * H6: batched read-modify-write for callers that need to patch many
 * runs at once (e.g. `staleRunReaper` flipping every old `running` run
 * to `failed` in one pass). Doing this with a loop of `updateRun` calls
 * costs N lock cycles, N reads, and N atomic renames; this helper does
 * 1 read + N in-memory patches + 1 write under a single lock, while
 * still firing the same per-run `transition` / `updated` events so
 * existing SSE consumers (the events route forwarding `transition` to
 * the UI) keep working.
 *
 * Patches whose `sessionId` doesn't match any run in `meta.runs` are
 * silently skipped — the reaper is the primary caller and a run that
 * was deleted between the meta read and the patch list is a no-op,
 * not an error. Returns the post-write `Meta` for the caller's
 * convenience (the reaper hands it straight to the API response).
 */
export async function applyManyRuns(
  dir: string,
  patches: Array<{ sessionId: string; patch: Partial<Run> }>,
): Promise<Meta | null> {
  if (patches.length === 0) {
    return readMeta(dir);
  }
  return withTaskLock(dir, () => {
    const meta = readMeta(dir);
    if (!meta) return null;
    // Buffer per-run events and flush them only AFTER the atomic write
    // so SSE subscribers never observe a `transition` event for a row
    // that isn't yet on disk.
    const pending: MetaChangeEvent[] = [];
    let mutated = false;
    for (const { sessionId, patch } of patches) {
      const run = meta.runs.find((r) => r.sessionId === sessionId);
      if (!run) continue;
      // Review nit: skip patches that are a no-op against the
      // current on-disk state. Primary case: the reaper's outer
      // `readMeta` saw the run as `running`, but between that read
      // and our locked re-read another writer flipped it to
      // `failed` (e.g. lifecycle hook firing). Without this guard
      // we'd emit a spurious `"updated"` SSE event and rewrite
      // identical bytes.
      const prevStatus = run.status;
      const statusUnchanged =
        patch.status === undefined || patch.status === prevStatus;
      const noOtherFields = Object.keys(patch).every(
        (k) => k === "status" || (run as unknown as Record<string, unknown>)[k] === (patch as Record<string, unknown>)[k],
      );
      if (statusUnchanged && noOtherFields) continue;
      Object.assign(run, patch);
      mutated = true;
      const statusChanged = !statusUnchanged;
      pending.push({
        taskId: taskIdFromDir(dir),
        kind: statusChanged ? "transition" : "updated",
        sessionId,
        run: { ...run },
        prevStatus,
      });
    }
    if (mutated) {
      atomicWriteJson(join(dir, FILE), meta);
      for (const ev of pending) emit(ev);
    }
    return meta;
  });
}

/**
 * H7: helper for the DELETE /api/sessions/[sessionId] handler — filter
 * a session out of the task's runs under the same lock that protects
 * appendRun/updateRun. Returns true if the session was found and
 * removed (so the caller can track which tasks were affected), false
 * if the meta is missing or the session wasn't linked to this task.
 */
export async function removeSessionFromTask(
  dir: string,
  sessionId: string,
): Promise<boolean> {
  return withTaskLock(dir, () => {
    const meta = readMeta(dir);
    if (!meta) return false;
    const before = meta.runs.length;
    meta.runs = meta.runs.filter((r) => r.sessionId !== sessionId);
    if (meta.runs.length === before) return false;
    atomicWriteJson(join(dir, FILE), meta);
    emit({ taskId: taskIdFromDir(dir), kind: "writeMeta" });
    return true;
  });
}

File: internal/meta/meta.go
// Package meta is the source of truth for per-task runtime state on
// disk: sessions/<task-id>/meta.json. Provides atomic write
// (tempfile + rename) with a per-directory mutex so concurrent
// AppendRun / UpdateRun callers serialize cleanly.
//
// Ported from libs/atomicWrite.ts + libs/tasks.ts + libs/meta.ts in S09.
// Event subscription (SubscribeMeta / SubscribeMetaAll) lives here too
// so the SSE route in S12 can subscribe without depending on the
// per-route singleton stash the TS module used.
package meta

import (
	"encoding/json"
	"errors"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
)

// RunStatus is the lifecycle state of one spawned claude session.
// Mirrors the TS string union: queued | running | done | failed | stale.
type RunStatus string

const (
	RunStatusQueued  RunStatus = "queued"
	RunStatusRunning RunStatus = "running"
	RunStatusDone    RunStatus = "done"
	RunStatusFailed  RunStatus = "failed"
	RunStatusStale   RunStatus = "stale"
)

// Run is one Claude Code session the coordinator spawned for a task.
// Field naming follows libs/meta.ts exactly so meta.json is bytewise
// interchangeable between the Next reader and the Go reader.
//
// Pointer-typed optional fields (e.g. ParentSessionID *string) keep
// json.Marshal emitting `null` rather than `""` when the field is
// absent — the TS code uses `null` for "explicitly absent" and the
// `omitempty` tag would lose that distinction.
//
// Verifier sub-types (Verify, Verifier, StyleCritic, SemanticVerifier)
// are kept as full Go structs even though S09's surface area doesn't
// touch them — preserving them as `json.RawMessage` would lose the
// shape contract and any future Go-side reader would have to re-derive
// it. The structs are declared in subtypes.go.
type Run struct {
	SessionID string    `json:"sessionId"`
	Role      string    `json:"role"`
	Repo      string    `json:"repo"`
	Status    RunStatus `json:"status"`
	StartedAt *string   `json:"startedAt"`
	EndedAt   *string   `json:"endedAt"`

	ParentSessionID *string `json:"parentSessionId,omitempty"`
	RetryOf         *string `json:"retryOf,omitempty"`
	RetryAttempt    *int    `json:"retryAttempt,omitempty"`

	Verify           *RunVerify           `json:"verify,omitempty"`
	Verifier         *RunVerifier         `json:"verifier,omitempty"`
	StyleCritic      *RunStyleCritic      `json:"styleCritic,omitempty"`
	SemanticVerifier *RunSemanticVerifier `json:"semanticVerifier,omitempty"`

	WorktreePath       *string `json:"worktreePath,omitempty"`
	WorktreeBranch     *string `json:"worktreeBranch,omitempty"`
	WorktreeBaseBranch *string `json:"worktreeBaseBranch,omitempty"`

	SpeculativeGroup   *string `json:"speculativeGroup,omitempty"`
	SpeculativeOutcome *string `json:"speculativeOutcome,omitempty"`
}

// Meta is the per-task runtime state. The full task definition (title,
// body, status, section, checked) lives here too — meta.json is the
// source of truth, replacing the old tasks.md round-trip.
//
// DetectedScope is held as json.RawMessage because its shape lives in
// the detect package (S16). meta.go round-trips the bytes verbatim so
// the eventual detect port doesn't need a coordinated cutover.
type Meta struct {
	TaskID        string          `json:"taskId"`
	TaskTitle     string          `json:"taskTitle"`
	TaskBody      string          `json:"taskBody"`
	TaskStatus    TaskStatus      `json:"taskStatus"`
	TaskSection   TaskSection     `json:"taskSection"`
	TaskChecked   bool            `json:"taskChecked"`
	TaskApp       *string         `json:"taskApp,omitempty"`
	CreatedAt     string          `json:"createdAt"`
	Runs          []Run           `json:"runs"`
	DetectedScope json.RawMessage `json:"detectedScope,omitempty"`
}

// MetaFile is the file basename inside each task's sessions dir.
const MetaFile = "meta.json"

// ErrMissingMeta is returned by AppendRun / UpdateRun when meta.json
// doesn't exist for the task dir. Distinct sentinel so callers can
// surface a 404 instead of a generic 500.
var ErrMissingMeta = errors.New("meta: meta.json missing")

// ErrRunNotFound is returned by UpdateRun when the named sessionID
// doesn't exist in meta.runs.
var ErrRunNotFound = errors.New("meta: run not found")

// ErrMetaExists is returned by CreateMeta when meta.json already
// exists in the target dir. Distinct from a generic write error so
// the API layer can return 409 Conflict instead of 500. Mirrors the
// "create-or-fail" semantics of POST /api/tasks where a duplicate
// task id must NOT clobber the existing meta.
var ErrMetaExists = errors.New("meta: meta.json already exists")

// ReadMeta loads meta.json from the task's sessions dir.
//
// Return values:
//   - (nil, nil) — meta.json doesn't exist (task missing / not yet
//     created). Callers map this to a 404.
//   - (nil, err) — read failure OR JSON parse failure. Parse errors
//     are surfaced rather than swallowed: a truncated file (operator
//     hand-edit, disk corruption, partial restore) signals fixable
//     damage and silently returning (nil, nil) made the dashboard
//     show "task missing" instead of telling the operator their
//     meta.json is broken. The parse failure is not cached.
//   - (m, nil) — happy path.
//
// Cached in-process via the package-global metaCache (TTL 500ms +
// 1024-entry LRU); cache invalidation happens on every emit() — so
// readers immediately see a writeMeta from another goroutine.
func ReadMeta(dir string) (*Meta, error) {
	if hit, ok := metaCache.get(dir); ok {
		return hit, nil
	}
	p := filepath.Join(dir, MetaFile)
	b, err := os.ReadFile(p)
	if err != nil {
		if errors.Is(err, fs.ErrNotExist) {
			metaCache.put(dir, nil)
			return nil, nil
		}
		return nil, fmt.Errorf("read %s: %w", p, err)
	}
	var m Meta
	if err := json.Unmarshal(b, &m); err != nil {
		// Don't poison the cache with the parse failure: next read after
		// a writeMeta will pick up the fresh bytes. Surface the error so
		// the operator/UI sees "meta.json corrupt" rather than the
		// misleading "task not found".
		return nil, fmt.Errorf("meta.json parse %s: %w", p, err)
	}
	metaCache.put(dir, &m)
	return &m, nil
}

// CreateMeta initializes a fresh meta.json. Caller passes the header
// fields; if Runs is nil it's initialized to an empty slice (the wire
// shape is `[]`, never `null`). A caller-provided non-nil Runs slice
// is preserved verbatim — useful for tests / restores that want to
// seed historical run rows in the same write.
//
// CreateMeta is "create-or-fail": if meta.json already exists in dir,
// returns ErrMetaExists rather than overwriting. Both the existence
// check and the write happen under the per-task lock, so two
// concurrent CreateMeta calls produce one success + one ErrMetaExists.
//
// Emits writeMeta on the package-global event bus so subscribers
// (S12 SSE route) see the create.
//
// Mirrors libs/meta.ts createMeta.
func CreateMeta(dir string, header Meta) error {
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return err
	}
	if header.Runs == nil {
		header.Runs = []Run{}
	}
	return WithTaskLock(dir, func() error {
		p := filepath.Join(dir, MetaFile)
		if _, err := os.Stat(p); err == nil {
			return ErrMetaExists
		} else if !errors.Is(err, fs.ErrNotExist) {
			return fmt.Errorf("stat %s: %w", p, err)
		}
		return WriteMetaUnlocked(dir, &header)
	})
}

// WriteMeta replaces meta.json on disk and notifies subscribers.
// Acquires the per-task lock, so callers MUST NOT already hold it
// (call WriteMetaUnlocked from inside WithTaskLock instead — or use
// MutateMeta for the read-modify-write pattern).
//
// Mirrors libs/meta.ts writeMeta — emits a writeMeta event (no per-run
// signal — task header changed).
func WriteMeta(dir string, m *Meta) error {
	return WithTaskLock(dir, func() error {
		return WriteMetaUnlocked(dir, m)
	})
}

// WriteMetaUnlocked is the lock-less worker for callers that already
// hold the per-task lock (AppendRun, UpdateRun,
// RemoveSessionFromTask, MutateMeta, the existence-checked tail of
// CreateMeta, and external API handlers that need to perform other
// operations under the same lock alongside the write). Writes the
// JSON, then emits the generic writeMeta event. Helpers that need a
// more specific event kind (spawned/transition/updated) should call
// WriteJSONAtomic + emit directly.
//
// Prefer MutateMeta for the read-modify-write pattern. WriteMeta auto-
// locks for the common case. Use WriteMetaUnlocked only when the
// caller already holds the lock via WithTaskLock.
func WriteMetaUnlocked(dir string, m *Meta) error {
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return err
	}
	if err := WriteJSONAtomic(filepath.Join(dir, MetaFile), m, nil); err != nil {
		return err
	}
	emit(dir, MetaChangeEvent{
		TaskID: taskIDFromDir(dir),
		Kind:   MetaChangeWriteMeta,
	})
	return nil
}

// MutateMeta runs fn(meta) under the per-task lock and writes the
// result back. fn receives the live *Meta loaded from disk; it may
// mutate fields in place. Returning a non-nil error from fn aborts
// the write and propagates up; returning nil triggers a write +
// writeMeta event.
//
// Sentinel returns:
//   - ErrMissingMeta — meta.json doesn't exist (mirrors AppendRun /
//     UpdateRun semantics so callers can branch on the same value).
//
// Group F's API handlers should prefer MutateMeta over the manual
// "WithTaskLock(dir, ReadMeta + WriteMeta)" sequence — it cuts the
// lock-acquisition boilerplate and centralises the missing-meta
// check.
func MutateMeta(dir string, fn func(*Meta) error) error {
	return WithTaskLock(dir, func() error {
		m, err := ReadMeta(dir)
		if err != nil {
			return err
		}
		if m == nil {
			return ErrMissingMeta
		}
		if err := fn(m); err != nil {
			return err
		}
		return WriteMetaUnlocked(dir, m)
	})
}

// AppendRun adds a Run to the task under the per-dir lock and emits a
// "spawned" event. Mirrors libs/meta.ts appendRun: the lock serializes
// the read window so back-to-back appends from concurrent goroutines
// can't lose-update each other.
func AppendRun(dir string, run Run) error {
	return WithTaskLock(dir, func() error {
		m, err := ReadMeta(dir)
		if err != nil {
			return err
		}
		if m == nil {
			return ErrMissingMeta
		}
		m.Runs = append(m.Runs, run)
		if err := WriteJSONAtomic(filepath.Join(dir, MetaFile), m, nil); err != nil {
			return err
		}
		emit(dir, MetaChangeEvent{
			TaskID:    taskIDFromDir(dir),
			Kind:      MetaChangeSpawned,
			SessionID: run.SessionID,
			Run:       &run,
		})
		return nil
	})
}

// UpdateRun applies patchFn to the run named by sessionID. patchFn
// mutates the Run in place (similar to TS Object.assign(run, patch));
// returning false from precondition skips the write entirely and
// returns (applied=false, nil). The optional precondition lets the
// caller reject demotions like done → failed under race.
//
// Mirrors libs/meta.ts updateRun. Emits a "transition" event when
// patchFn changes Status, "updated" otherwise.
func UpdateRun(dir, sessionID string, patchFn func(*Run), precondition func(Run) bool) (applied bool, err error) {
	err = WithTaskLock(dir, func() error {
		m, rerr := ReadMeta(dir)
		if rerr != nil {
			return rerr
		}
		if m == nil {
			return ErrMissingMeta
		}
		idx := -1
		for i, r := range m.Runs {
			if r.SessionID == sessionID {
				idx = i
				break
			}
		}
		if idx < 0 {
			return ErrRunNotFound
		}
		if precondition != nil && !precondition(m.Runs[idx]) {
			applied = false
			return nil
		}
		prevStatus := m.Runs[idx].Status
		patchFn(&m.Runs[idx])
		if werr := WriteJSONAtomic(filepath.Join(dir, MetaFile), m, nil); werr != nil {
			return werr
		}
		applied = true
		kind := MetaChangeUpdated
		if m.Runs[idx].Status != prevStatus {
			kind = MetaChangeTransition
		}
		runCopy := m.Runs[idx]
		emit(dir, MetaChangeEvent{
			TaskID:     taskIDFromDir(dir),
			Kind:       kind,
			SessionID:  sessionID,
			Run:        &runCopy,
			PrevStatus: prevStatus,
		})
		return nil
	})
	return
}

// RemoveSessionFromTask is the helper for the DELETE /api/sessions/<id>
// handler — filters a session out of the task's runs under the same
// lock that protects appendRun/updateRun. Returns true when the
// session was found + removed, false when meta is missing or the
// session wasn't linked to this task.
//
// Mirrors libs/meta.ts removeSessionFromTask.
func RemoveSessionFromTask(dir, sessionID string) (removed bool, err error) {
	err = WithTaskLock(dir, func() error {
		m, rerr := ReadMeta(dir)
		if rerr != nil {
			return rerr
		}
		if m == nil {
			return nil
		}
		before := len(m.Runs)
		// Filter-in-place: out aliases the same backing array as
		// m.Runs, but the value-copy `for _, r := range m.Runs`
		// snapshots each Run before the conditional append, so
		// overwriting earlier slots while iterating later ones is
		// safe. The post-loop `m.Runs = out` shortens the slice
		// header to the kept count.
		out := m.Runs[:0]
		for _, r := range m.Runs {
			if r.SessionID == sessionID {
				continue
			}
			out = append(out, r)
		}
		m.Runs = out
		if len(m.Runs) == before {
			return nil
		}
		if werr := WriteJSONAtomic(filepath.Join(dir, MetaFile), m, nil); werr != nil {
			return werr
		}
		emit(dir, MetaChangeEvent{
			TaskID: taskIDFromDir(dir),
			Kind:   MetaChangeWriteMeta,
		})
		removed = true
		return nil
	})
	return
}

// taskIDFromDir mirrors basename(dir) — every callsite uses
// filepath.Join(SESSIONS_DIR, taskID), so basename is the task id.
func taskIDFromDir(dir string) string {
	return filepath.Base(dir)
}
